File: Dimensionality_Reduction/LDA/lda.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Linear Discriminant Analysis (LDA) - Dimensionality Reduction
"""

# Importing necessary libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import os
from sklearn.preprocessing import StandardScaler
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis as LDA
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix
import seaborn as sns

# Get the directory where the script is located
script_dir = os.path.dirname(os.path.abspath(__file__))
file_path = os.path.join(script_dir, "Wine.csv")  # Dataset file

# Checking if the dataset exists
if os.path.isfile(file_path):
    dataset = pd.read_csv(file_path)
    print("Data imported successfully!")  
else:
    raise FileNotFoundError(f"Error: File not found: {file_path}. Please make sure 'Wine.csv' is in the same directory.")

# Splitting dataset into features (X) and target (y)
X = dataset.iloc[:, 0:13].values  # Selecting first 13 columns (Chemical properties)
y = dataset.iloc[:, 13].values    # Target column (Wine class)

# Splitting the dataset into training and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=0)

# Feature Scaling (Required for LDA)
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

# Applying LDA - Choosing top 2 Linear Discriminants
lda = LDA(n_components=2)
X_train_lda = lda.fit_transform(X_train, y_train)
X_test_lda = lda.transform(X_test)

# Training Logistic Regression on LDA-transformed data
classifier = LogisticRegression(random_state=0)
classifier.fit(X_train_lda, y_train)

# Predicting the test set results
y_pred = classifier.predict(X_test_lda)

# Evaluating model performance
accuracy = accuracy_score(y_test, y_pred)
cm = confusion_matrix(y_test, y_pred)

# Printing the results
print(f"Model Accuracy after LDA: {accuracy * 100:.2f}%")
print("Confusion Matrix:")
print(cm)

# Plotting confusion matrix using Seaborn
plt.figure(figsize=(5, 4))
sns.heatmap(cm, annot=True, fmt="d", cmap="Blues", xticklabels=['Class 0', 'Class 1', 'Class 2'], yticklabels=['Class 0', 'Class 1', 'Class 2'])
plt.title('Confusion Matrix (After LDA)')
plt.xlabel('Predicted')
plt.ylabel('Actual')
plt.show()

# Visualizing the LDA results (Training Set)
plt.figure(figsize=(7,5))
plt.scatter(X_train_lda[:, 0], X_train_lda[:, 1], c=y_train, cmap="coolwarm", edgecolor="k")
plt.title("LDA Applied to Wine Dataset (Training Set)")
plt.xlabel("Linear Discriminant 1 (LD1)")
plt.ylabel("Linear Discriminant 2 (LD2)")
plt.colorbar()
plt.show()

File: Dimensionality_Reduction/LDA/LDA.R
# LDA (Linear Discriminant Analysis)

# Importing Dataset
dataset <- read.csv('Wine.csv')

# Splitting Data into Training and Testing Sets
library(caTools)
set.seed(123)  # Ensures reproducibility
split <- sample.split(dataset$Customer_Segment, SplitRatio = 0.80)
training_set <- subset(dataset, split == TRUE)
test_set <- subset(dataset, split == FALSE)

# Feature Scaling (Apply Only to Numeric Columns)
training_set[, -14] <- scale(training_set[, -14])
test_set[, -14] <- scale(test_set[, -14])

# Applying LDA
library(MASS)
lda_model = lda(formula = Customer_Segment ~ ., data = training_set)
training_set_lda = as.data.frame(predict(lda_model, training_set))
training_set_lda = training_set_lda[c("x.LD1", "x.LD2", "class")]

test_set_lda = as.data.frame(predict(lda_model, test_set))
test_set_lda = test_set_lda[c("x.LD1", "x.LD2", "class")]

# Convert class labels back to factor
training_set_lda$class <- as.factor(training_set_lda$class)
test_set_lda$class <- as.factor(test_set_lda$class)

# Fitting SVM to the Training set
library(e1071)
classifier = svm(formula = class ~ .,
                 data = training_set_lda,
                 type = 'C-classification',
                 kernel = 'linear')

# Predicting the Test Set Results
y_pred = predict(classifier, newdata = test_set_lda[-3])

# Creating the Confusion Matrix
cm <- table(test_set_lda$class, y_pred)
print(cm)

# Visualization Function for Training and Test Sets
visualize_results <- function(set, title, classifier) {
  X1 = seq(min(set[, 1]) - 1, max(set[, 1]) + 1, by = 0.01)
  X2 = seq(min(set[, 2]) - 1, max(set[, 2]) + 1, by = 0.01)
  grid_set = expand.grid(X1, X2)
  colnames(grid_set) = c('LD1', 'LD2')
  
  y_grid = predict(classifier, newdata = grid_set)
  
  plot(set[, -3],
       main = title,
       xlab = 'Linear Discriminant 1', ylab = 'Linear Discriminant 2',
       xlim = range(X1), ylim = range(X2))
  contour(X1, X2, matrix(as.numeric(y_grid), length(X1), length(X2)), add = TRUE)
  
  points(grid_set, pch = '.', col = ifelse(y_grid == 2, 'deepskyblue', 
                                           ifelse(y_grid == 1, 'springgreen', 'tomato')))
  points(set, pch = 21, bg = ifelse(set[, 3] == 2, 'blue3', 
                                    ifelse(set[, 3] == 1, 'green4', 'red3')))
}

# Visualizing the Training Set Results
visualize_results(training_set_lda, "SVM (Training Set with LDA)", classifier)

# Visualizing the Test Set Results
visualize_results(test_set_lda, "SVM (Test Set with LDA)", classifier)
